### PR TITLE
Initializer function loosing context issue

### PR DESCRIPTION
**Pull request by JoseLion**
@aexmachina during some work with the library I noticed that the sequence function was not increasing when used inside of an initializer function:
```javascript
Factory.define('FucSeq', DummyModel, options => ({
  s1: Factory.seq(),
  s2: Factory.seq('FuncSeq.s2'),
  s3: Factory.seq(n => `${options.prefix}-${n}`)
}));
```
In the code above only `s2` will work because we are passing an ID to the generator, but `s1` and `s3` will fail to increase the sequence correctly, it seems a new Factory instance is created each time (not really sure why).

I found there is a test for this but it was incomplete, it was only covering the case of `s2` (please check `test/test-helper/dummyFactories.js:46`), so I added the missing cases for `s1` and `s3` and attempted to fix the issue, which unfortunately proved to be a difficult task.

In this PR I tried to solve the issue by resolving the initializer function before it looses context of the Factory (in the `define` function of `src/FactoryGirl.js`), but obviously this doesn't fully works since we loose the options object that may be passed in the initializer function.

I thought I could seek your advice on how we could solve this issue in the best way, so I hope the explanation is clear enough :)

Thank you in advance for your time!

### CODE REVIEW

File: test/test-helper/dummyFactories.js [hunk 1]
```diff
@@ -39,9 +39,11 @@ Factory.define('User', User, {
 Factory.define('ObjSeq', DummyModel, {
   s1: Factory.seq(),
   s2: Factory.seq('ObjSeq.s2'),
-  s3: Factory.seq(),
+  s3: Factory.seq(n => n),
```
Comment (simonexmachina) [inline test/test-helper/dummyFactories.js]: Why is this required, isn't this default behaviour?

**Review by simonexmachina — COMMENTED** [review]
I'm not sure I understand the problem you're referring to. Is there a failing test? I'm not really doing code much any more so I'm afraid that I'm not the person to help work out the issue. @chetanism are you able to assist @JoseLion with this?

**Comment by JoseLion** [comment]
@aexmachina sorry took so long to answer. I'll post a code snippet that shows the issue so you can easily test it in [katacoda](https://www.katacoda.com/courses/nodejs/playground). @chetanism I would really appreciate the help if you have some time to spare :)

This first snippet runs the sequences defining the `user` factory with the initializer passed as an object and everything goes as expected:
```javascript
const factory = require('factory-girl').factory;

class User {
  constructor({username, score}) {
    this.username = username;
    this.score = score;
  }
}

const user = {
  name: 'Bob',
  finalScore: 0
};

factory.define('user', User, {
  username: user.name,
  score: factory.sequence(n => `${n} pts.`)
});

factory.build('user').then(user => {
  console.log(user); // => User {username: 'Bob', score: '1 pts.'}
});
factory.build('user').then(user => {
  console.log(user); // => User {username: 'Bob', score: '2 pts.'}
});
factory.build('user').then(user => {
  console.log(user); // => User {username: 'Bob', score: '3 pts.'}
});
```

This other snippet does the exact same thing but the initializer is now a function, so the suffix may be passed dynamically to the sequence generator:
```javascript
const factory = require('factory-girl').factory;

class User {
  constructor({username, score}) {
    this.username = username;
    this.score = score;
  }
}

const user = {
  name: 'Bob',
  finalScore: 0
};

factory.define('user', User, options => ({
  username: user.name,
  score: factory.sequence(n => `${n} ${options.suffix}`)
}));

factory.build('user', {}, { suffix: "pts." }).then(user => {
  console.log(user); // => User {username: 'Bob', score: '1 pts.'}
});
factory.build('user', {}, { suffix: "pts." }).then(user => {
  console.log(user); // => User {username: 'Bob', score: '1 pts.'}
});
factory.build('user', {}, { suffix: "pts." }).then(user => {
  console.log(user); // => User {username: 'Bob', score: '1 pts.'}
});
```
As you can see the sequence generator fails to keep adding to numbers to the sequence instance, apparently a new instance of `Factory` is created each time the `factory.build(...)` is called. I've created the test cases for this issue to have somewhere to start testing, but I couldn't find a complete solution.

I would really appreciate any help I can get, I'll be happy to contribute to the library by fixing this :)

File: test/test-helper/dummyFactories.js (hunk 1 above)
Comment (JoseLion) [reply test/test-helper/dummyFactories.js]: @aexmachina you're right this would be the default behaviour. The idea is to test that passing a function as initializer would work as expected, so I'll change to something better 👍
